### PR TITLE
[MRG] dt and t checks

### PR DESCRIPTION
**Pull request by mstimberg**
This should implement what we discussed in #730. Let me know if I forgot anything or misunderstood something.

### CODE REVIEW

File: brian2/core/network.py [hunk 1]
```diff
@@ -715,12 +715,14 @@ def after_run(self):
...
+                            self._clock_variables[c][1][0],
+                            self._clock_variables[c][2][0])
                         for c in self._clocks]
-        minclock, min_time = min(clocks_times, key=lambda k: k[1])
-        curclocks = set(clock for clock, time in clocks_times if
+        minclock, min_time, minclock_dt = min(clocks_times_dt, key=lambda k: k[1])
+        curclocks = set(clock for clock, time, dt in clocks_times_dt if
                         (time == min_time or
-                         abs(time - min_time) < Clock.epsilon))
+                         abs(time - min_time)/min([minclock_dt, dt]) < Clock.epsilon_dt))
```
Comment (thesamovar) [inline brian2/core/network.py]: Mildly more efficient not to use a list for ``[minclock_dt, dt]`` but just pass them as arguments directly? Not sure actually.

**Comment by thesamovar** [comment]
The ``check_dt`` function is a bit strange because there's an ``old_dt`` and ``new_dt`` but also ``target_t``. If it's changing from one dt to another dt, why not just directly pass the ``old_i`` value rather than the ``target_t`` value? That seems safer to me, rather than inferring the ``old_i`` value from ``target_t``. Maybe it's OK because any time the epsilon check is passed, ``round`` would give that same integer value.

File: brian2/core/network.py (hunk 1 above)
Comment (mstimberg) [inline brian2/core/network.py]: You are right, it is a bit more efficient (a tuple would be fine but a list takes more time) -- I think I had numpy's min in mind where you cannot use that syntax.

**Comment by mstimberg** [comment]
> The check_dt function is a bit strange because there's an old_dt and new_dt but also target_t. If it's changing from one dt to another dt, why not just directly pass the old_i value rather than the target_t value? That seems safer to me, rather than inferring the old_i value from target_t. Maybe it's OK because any time the epsilon check is passed, round would give that same integer value.

The whole thing is not very straightforward, the function where `check_dt` is called does not only set `dt`, but also `t`. We delay the actual change of `dt` until the start of the simulation, this is because otherwise code like this would throw an error:
```Python
group = ...
net1 = Network(group)
net1.run(0.3*ms)
defaultclock.dt = 0.5*ms  # this is not ok if we continue the run, but...
group = ...
net2 = Network(group)
net2.run(1*ms)  #... we actually start at 0 again
```
For the second run, we set both `t` and `dt` at once, we don't have a previous time step. I am sure one could write this differently, but the whole system is rather fragile and I'd prefer not to make changes if not necessary.

**Comment by thesamovar** [comment]
Ah OK, I understand now. In that case, I think it's good to go! Will you make the minor change to ``max``? Probably doesn't make much difference but may as well have it I guess.

**Comment by mstimberg** [comment]
I did the change and also added a warning for extremely long simulations which I forgot. All good to go from my side now.

**Comment by thesamovar** [comment]
Great! Good to merge when it passes then.

**Comment by mstimberg** [comment]
Oh, forgot to merge this. Merging it now

**Pull request merged**